shared/src/uniswap_pool.rs

```diff
@@ -138,6 +143,64 @@ impl Pool {
     }
 }
 
+// Read though Pool Fetcher that keeps previously fetched pools in a cache, which get invalidated whenever there is a new block
+pub struct CachedPoolFetcher<S> {
+    inner: Box<dyn PoolFetching>,
+    cache: Arc<Mutex<HashMap<TokenPair, Pool>>>,
+    block_stream: Arc<Mutex<S>>,
+}
+
+impl<S> CachedPoolFetcher<S> {
+    pub fn new(inner: Box<dyn PoolFetching>, block_stream: S) -> Self {
+        Self {
+            inner,
+            cache: Arc::new(Mutex::new(HashMap::new())),
+            block_stream: Arc::new(Mutex::new(block_stream)),
+        }
+    }
+
+    async fn fetch_inner(&self, token_pairs: HashSet<TokenPair>) -> Vec<Pool> {
+        let mut cache = self.cache.lock().await;
+        let (cache_hits, cache_misses) = token_pairs
+            .into_iter()
+            .partition::<HashSet<_>, _>(|pair| cache.contains_key(pair));
+        let cache_results: Vec<_> = cache_hits
+            .iter()
+            .filter_map(|pair| cache.get(pair))
+            .cloned()
+            .collect();
+
+        let mut inner_results = self.inner.fetch(cache_misses).await;
+        for miss in &inner_results {
+            cache.insert(miss.tokens, miss.clone());
+        }
+        inner_results.extend(cache_results);
+        inner_results
+    }
+
+    async fn clear_cache(&self) {
+        let mut cache = self.cache.lock().await;
+        cache.clear();
+    }
+}
+
+#[async_trait::async_trait]
+impl<S: Stream<Item = Block> + FusedStream + Send + Sync + Unpin> PoolFetching
+    for CachedPoolFetcher<S>
+{
+    async fn fetch(&self, token_pairs: HashSet<TokenPair>) -> Vec<Pool> {
+        let fetch = || self.fetch_inner(token_pairs);
+        let clear_cache = || self.clear_cache();
+        let mut stream = self.block_stream.lock().await;
+        loop {
+            select! {
+                _ = stream.next() => clear_cache().await,
+                default => return fetch().await,
+            };
+        }
```

Comment:
If I understand this right this is supposed to clear the cache if stream is immediately ready and otherwise return `fetch()`. It might be possible to express this more obviously with `FutureExt::now_or_never` instead of `select!`.